zookeeper-server/src/test/java/org/apache/zookeeper/server/quorum/UnifiedServerSocketTest.java

```diff
@@ -17,156 +17,644 @@
+        UnifiedServerThread serverThread = new UnifiedServerThread(
+                x509Util, localServerAddress, true, DATA_TO_CLIENT);
+        serverThread.start();
+
+        Socket socket = connectWithoutSSL();
+        // Write only 2 bytes of the message, wait a bit, then write the rest.
+        // This makes sure that writes smaller than 5 bytes don't break the plaintext mode on the server
+        // once it decides that the input doesn't look like a TLS handshake.
+        socket.getOutputStream().write(DATA_FROM_CLIENT, 0, 2);
+        socket.getOutputStream().flush();
+        Thread.sleep(TIMEOUT / 2);
+        socket.getOutputStream().write(DATA_FROM_CLIENT, 2, DATA_FROM_CLIENT.length - 2);
+        socket.getOutputStream().flush();
+        byte[] buf = new byte[DATA_TO_CLIENT.length];
+        int bytesRead = socket.getInputStream().read(buf, 0, buf.length);
+        Assert.assertEquals(buf.length, bytesRead);
+        Assert.assertArrayEquals(DATA_TO_CLIENT, buf);
+
+        serverThread.shutdown(TIMEOUT);
+        forceClose(socket);
 
-        class ServerThread extends Thread {
+        Assert.assertArrayEquals(DATA_FROM_CLIENT, serverThread.getDataFromClient(0));
+    }
+
+    /**
+     * Attempting to connect to a SSL-only server without SSL should fail.
+     */
+    @Test
+    public void testConnectWithoutSSLToStrictServer() throws Exception {
+        UnifiedServerThread serverThread = new UnifiedServerThread(
+                x509Util, localServerAddress, false, DATA_TO_CLIENT);
+        serverThread.start();
+
+        Socket socket = connectWithoutSSL();
+        socket.getOutputStream().write(DATA_FROM_CLIENT);
+        socket.getOutputStream().flush();
+        byte[] buf = new byte[DATA_TO_CLIENT.length];
+        try {
+            socket.getInputStream().read(buf, 0, buf.length);
+        } catch (SocketException e) {
+            // We expect the other end to hang up the connection
+            return;
+        } finally {
+            serverThread.shutdown(TIMEOUT);
+            forceClose(socket);
+        }
+        Assert.fail("Expected server to hang up the connection. Read from server succeeded unexpectedly.");
+    }
+
+    /**
+     * This test makes sure that UnifiedServerSocket used properly (a single thread accept()-ing connections and
+     * handing the resulting sockets to other threads for processing) is not vulnerable to a simple denial-of-service
+     * attack in which a client connects and never writes any bytes. This should not block the accepting thread, since
+     * the read to determine if the client is sending a TLS handshake or not happens in the processing thread.
+     *
+     * This version of the test uses a non-strict server socket (i.e. it accepts both TLS and plaintext connections).
+     */
+    @Test
+    public void testDenialOfServiceResistanceNonStrictServer() throws Exception {
+        UnifiedServerThread serverThread = new UnifiedServerThread(
```

Comment:
In these DOS tests you're actually testing the `UnifiedServerThread` implementation which is a test-only thing. Does it make sense or am I missing something here?

In "reality" the accept thread is basically the `Leader` and the one which does TLS handshake is the `LearnerHandler`.

Reply:
This is true. I tried to make the threading behavior in this test as similar as possible to `Leader`. I don't think Leader should be involved in this unit test, but perhaps a similar "denial of service resistance" test case can be added to some quorum test. Do you know which unit test would be a good place for such a test case?

Reply:
I would create new test class inherited from `QuorumPeerTestBase` dedicated to DOS tests, but I think this topic leads to another: stress testing ZooKeeper. I highly support the effort, but please remove them from this PR and create separate Jira for that.

Reply:
How about we keep the unit test that's focused strictly on UnifiedServerSocket in this PR, and I'll open a separate PR/Jira for adding a new unit test that tests `Leader`? There is some value in the existing tests, since they did catch issues with the original UnifiedServerSocket implementation.

Reply:
@ivmaykov That's pretty much the same as I suggested: keep all unit tests except the ones which contains the word "dos". Also please open a new jira about adding DOS tests.